### ISSUE TITLE
BUG: XGBRFRegressor invalid number of trees error with shap>=0.42

### ISSUE THREAD

**Issue by DawidDabkowski** [shap/shap#3484]
### Issue Description

`XGBRFRegressor` from `xgboost` raises "Invalid number of trees" error `shap>=0.42` when running `shap_values` method from `shap.TreeExplainer`. I tried it on Windows 10 with conda environment on Python 3.8.18 and on Linux environment with Python 3.8.11. Minimal install:
```
pip install shap==0.44.1
pip install xgboost==1.7.6
```

I checked that with `shap<0.42` there is no such issue. Also with `shap>=0.42` similar model `XGBRFClassifier` works fine, the problem seem to be only with `XGBRFRegressor`.

### Minimal Reproducible Example

```python
import numpy as np
from xgboost import XGBRFRegressor
from shap import TreeExplainer

num_points = 50
num_features = 5
rng = np.random.default_rng(seed=1)
X = rng.normal(size=(num_points, num_features))
y = rng.normal(size=(num_points,))
model = XGBRFRegressor().fit(X, y)
explainer = TreeExplainer(model, feature_perturbation="tree_path_dependent", model_output="raw")
explainer.shap_values(X)
```

### Traceback

```shell
Traceback (most recent call last):
  File "C:\Users\...\xgb_rf_regressor_test.py", line 12, in <module>
    explainer.shap_values(X)
  File "C:\Users\...\AppData\Local\miniconda3\envs\test_env\lib\site-packages\shap\explainers\_tree.py", line 418, in shap_values
    model_output_vals = self.model.original_model.predict(
  File "C:\Users\...\AppData\Local\miniconda3\envs\test_env\lib\site-packages\xgboost\core.py", line 2163, in predict
    _check_call(
  File "C:\Users\...\AppData\Local\miniconda3\envs\test_env\lib\site-packages\xgboost\core.py", line 279, in _check_call
    raise XGBoostError(py_str(_LIB.XGBGetLastError()))
xgboost.core.XGBoostError: [13:16:44] C:\buildkite-agent\builds\buildkite-windows-cpu-autoscaling-group-i-0fdc6d574b9c0d168-1\xgboost\xgboost-ci-windows\src\gbm\gbtree.cc:549: Check failed: tree_end <= model_.trees.size() (10000 vs. 100) : Invalid number of trees.

Process finished with exit code 1
```

### Bug report checklist

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest release](https://github.com/shap/shap/releases) of shap.
- [ ] I have confirmed this bug exists on the [master branch](https://github.com/shap/shap/blob/master/CONTRIBUTING.md#installing-from-the-master-branch) of shap.
- [ ] I'd be interested in making a PR to fix this bug

### Installed Versions

0.44.1

**Comment by connortann** [comment]
Thanks for the bug report!

On the master branch of shap, I can reproduce the issue above with `xgboost == 1.7.6` but not with `xgboost == 2.0.0`.

EDIT: It looks like a difference in the loaded `tree_limit`:

https://github.com/shap/shap/blob/bf2e44920b85d4f1173d42deff01ec7e6429bd47/shap/explainers/_tree.py#L389-L390

This is then passed to `iteration_range`, which replaces the deprecated xgboost parameter `ntree_limit` as of #2987 :

https://github.com/shap/shap/blob/bf2e44920b85d4f1173d42deff01ec7e6429bd47/shap/explainers/_tree.py#L402-L408

- with XGBoost `1.7.6`, `self.model.tree_limit` is `100`. The original model is called with `interation_range=(0, 100)`.
- with XGBoost `2.0.0`, `self.model.tree_limit` is `None`. The original model is called with `iteration_range=(0, 0)`.

The RFRegressor fails with an iteration_range that is higher than 1, presumably because the trees are in parallel. So, I think `iteration_range` is not quite a drop-in replacement for `ntree_limit`. Perhaps the `interation_range` should be something like `tree_limit / num_parallel_trees`?

On a side note, it doesn't look like there are any tests for `XGBRFRegressor` or `XGBRFClassifier` at the moment, so it would be good to explore if shap can officially support them. @trivialfis , do you by any chance know about this history of XGBoost & shap, if there is any particular reason why the XGB RF models are seemingly not already supported / tested by shap?

**Comment by connortann** [comment]
This may be solved by #3462 , which will improve the XGBoost model loading significantly. As a temporary workaround for now, you could try updating to XGBoost 2+ if possible.

**Comment by connortann** [comment]
This should be solved by #3462, and will be released in `0.45.0`.
